chia/full_node/full_node.py

```diff
@@ -1257,23 +1279,35 @@ async def add_block_batch(
         all_blocks: List[FullBlock],
         peer_info: PeerInfo,
         fork_info: Optional[ForkInfo],
+        current_ssi: uint64,
+        current_difficulty: uint64,
+        prev_ses_block: Optional[BlockRecord] = None,
         wp_summaries: Optional[List[SubEpochSummary]] = None,
-    ) -> Tuple[bool, Optional[StateChangeSummary], Optional[Err]]:
+    ) -> Tuple[bool, Optional[StateChangeSummary], uint64, uint64, Optional[BlockRecord], Optional[Err]]:
         # Precondition: All blocks must be contiguous blocks, index i+1 must be the parent of index i
         # Returns a bool for success, as well as a StateChangeSummary if the peak was advanced
+
+        assert current_ssi is not None and current_difficulty is not None
         block_dict: Dict[bytes32, FullBlock] = {}
         for block in all_blocks:
             block_dict[block.header_hash] = block
 
         blocks_to_validate: List[FullBlock] = []
         for i, block in enumerate(all_blocks):
             header_hash = block.header_hash
-            rec = await self.blockchain.get_block_record_from_db(header_hash)
-            if rec is not None:
-                self.blockchain.add_block_record(rec)
-            else:
+            block_rec = await self.blockchain.get_block_record_from_db(header_hash)
+            if block_rec is None:
                 blocks_to_validate = all_blocks[i:]
                 break
+            else:
+                self.blockchain.add_block_record(block_rec)
+                if block_rec.sub_epoch_summary_included:
+                    # already validated block, update sub slot iters, difficulty and prev sub epoch summary
+                    prev_ses_block = block_rec
+                    if block_rec.sub_epoch_summary_included.new_sub_slot_iters is not None:
+                        current_ssi = block_rec.sub_epoch_summary_included.new_sub_slot_iters
+                    if block_rec.sub_epoch_summary_included.new_difficulty is not None:
+                        current_difficulty = block_rec.sub_epoch_summary_included.new_difficulty
```

Comment:
here we update `current_ssi`, `current_difficulty` and `prev_ses_block` as we iterate through the blocks we're about to validate and add.
After this block of code, these variables are ready to be returned and then passed in to the subsequent call to `add_block_batch()`.

However, further down in this function, in the block around line 1390, we update these *again*, from what I can tell, in the same way, while iterating over `blocks_to_validate`.

Would it make sense to, at this point, define new variables that are updated, e.g. `new_ssi`, `new_difficulty` and `new_ses_block` and have them be immutable and return these new values unconditionally. I understand that we still need to "scan" the blocks again as we add the blocks again. But I think it would make it a bit easier to read by just advance the input variables over the blocks *once*.

If we end up updating these here, we don't regress them anywhere when we start over again, do we?
It seems like this could cause errors.

Reply:
here we update because we already validated these blocks (they where in the db) and we are not going to validate them this is to keep track of the changes up to the blocks we do validate

in 1390 we update as we actually do the valiation, 

it would not make sense to give different names because it is the current ssi/diff/ses block refering to where we are in the validation